src/features.c

```diff
@@ -0,0 +1,191 @@
+/*
+ * crun - OCI runtime written in C
+ *
+ * crun is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * crun is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with crun.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <config.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <argp.h>
+#include <string.h>
+#include <unistd.h>
+#include <errno.h>
+
+#include <yajl/yajl_tree.h>
+#include <yajl/yajl_gen.h>
+
+#include "crun.h"
+#include "libcrun/container.h"
+#include "libcrun/utils.h"
+
+static char doc[] = "OCI runtime";
+
+static struct argp_option options[] = { {0} };
+
+static char args_doc[] = "features";
+
+static error_t
+parse_opt (int key, char *arg arg_unused, struct argp_state *state arg_unused)
+{
+  if (key != ARGP_KEY_NO_ARGS) {
+        return ARGP_ERR_UNKNOWN;
+    }
+
+    return 0;
+}
+
+static struct argp run_argp = { options, parse_opt, args_doc, doc, NULL, NULL, NULL };
+
+int crun_command_features(struct crun_global_arguments *global_args, int argc, char **argv, libcrun_error_t *err) {
+    argp_parse(&run_argp, argc, argv, 0, 0, &options);
+
+    // Call the function in features.c to gather the feature information
+    struct features_info_s info;
+    int result = libcrun_container_print_features(&info);
+    if (result != 0) {
+        libcrun_make_error(err, result, "Failed to gather features information.");
+        return result;
+    }
+
+    // Prepare the JSON output
+    yajl_gen jsonGen = yajl_gen_alloc(NULL);
+    yajl_gen_status status;
+
+    yajl_gen_config(jsonGen, yajl_gen_beautify, 1); // Optional: Enable pretty formatting
+
+    // Start building the JSON
+    yajl_gen_map_open(jsonGen);
+
+    // Add ociVersionMin field
+    yajl_gen_string(jsonGen, (const unsigned char *) "ociVersionMin", strlen("ociVersionMin"));
+    yajl_gen_string(jsonGen, (const unsigned char *) info.ociVersionMin, strlen(info.ociVersionMin));
+
+    // Add ociVersionMax field
+    yajl_gen_string(jsonGen, (const unsigned char *) "ociVersionMax", strlen("ociVersionMax"));
+    yajl_gen_string(jsonGen, (const unsigned char *) info.ociVersionMax, strlen(info.ociVersionMax));
+
+    // Add hooks array
```

Comment:
this is a personal preference thing and is kind of a nit: it's tough to tell what the structure will look like and whether each open has a corresponding close based on just looking at this. I think it would be easier to read if each section had a function that corresponded to the fields required. For instance

```
crun_features_add_hooks(...)
{
    yajl_gen_string(jsonGen, (const unsigned char *) "hooks", strlen("hooks"));
    yajl_gen_array_open(jsonGen);

    for (size_t i = 0; info.hooks[i] != NULL; i++) {
        yajl_gen_string(jsonGen, (const unsigned char *) info.hooks[i], strlen(info.hooks[i]));
    }

    yajl_gen_array_close(jsonGen);
}
```
this way, you can tell just by looking at the function that an open corresponds to a close, and what contents are in that array.

up to you and the crun maintainers though

Reply:
I agree, let's make it clearer where the blocks start and end.

Reply:
ok, I fixed this in the latest push